torch/csrc/api/include/torch/data/datasets/chunk.h

```diff
@@ -0,0 +1,432 @@
+struct UnwrappedBatchData {
+ public:
+  using UnwrappedBatchType = UnwrappedBatch;
+
+  UnwrappedBatchData(UnwrappedBatchType data) : batch_data(std::move(data)) {}
+
+  UnwrappedBatchData(std::exception_ptr e) : exception(e) {}
+
+  /// batch data to return
+  UnwrappedBatchType batch_data;
+
+  /// exception pointer which captures any abnormal exceptions while creating the
+  /// batch.
+  std::exception_ptr exception;
+};
+
+/// BatchDataBuffer manages a queue of UnwrappedBatchData. After a new chunk is
+/// loaded, BatchDataBuffer splits it into small batches and push them into the
+/// queue. When get_batch is called from data loader, it pops cached batches and
+/// return. If the cache is empty, it either waits to load more chunks or return
+/// null if all chunks are loaded.
+template <
+    typename UnwrappedBatch = std::vector<Example<>>,
+    typename ExampleSampler = samplers::RandomSampler>
+class BatchDataBuffer {
+ public:
+  using UnwrappedBatchType = UnwrappedBatch;
+  using BatchType = torch::optional<UnwrappedBatchType>;
+  using BatchRequestType = typename ExampleSampler::BatchRequestType;
+
+  BatchDataBuffer(
+      size_t num_chunks,
+      size_t batch_size,
+      ExampleSampler example_sampler,
+      bool ignore_empty_chunk,
+      size_t cache_size)
+      : remaining_chunk_count_(num_chunks),
+        batch_size_(batch_size),
+        example_sampler_(std::move(example_sampler)),
+        ignore_empty_chunk_(ignore_empty_chunk),
+        queue_depth_(cache_size) {}
+
+  /// Return batch data from the queue. Called from the ChunkDataSet main
+  /// thread.
+  BatchType get_batch(size_t batch_size) {
+    std::unique_lock<std::mutex> lock(queue_mutex_);
+    cvr_.wait(lock, [this] {
+      // wait till there is available data in the queue or if all chunks are
+      // loaded (i.e. the data set is exhausted for this epoch)
+      return (
+          this->total_example_count_in_queue_ >= batch_size_ ||
+          remaining_chunk_count_ == 0);
+    });
+    if (batch_queue_.empty()) {
+      lock.unlock();
+      AT_ASSERT(remaining_chunk_count_ == 0);
+
+      // All batches have been retrieved. Return an empty batch.
+      return nullopt;
+    }
+
+    auto batch_data = batch_queue_.front();
+    batch_queue_.pop();
+    if (batch_data.exception) {
+      throw WorkerException(batch_data.exception);
+    }
+
+    this->total_example_count_in_queue_ -= batch_data.batch_data.size();
+    lock.unlock();
+    cvw_.notify_all(); // notify all writers.
+
+    return batch_data.batch_data;
+  }
+
+  // skip one chunk
+  void skip_chunk() {
+    std::unique_lock<std::mutex> lock(queue_mutex_);
+    AT_ASSERT(remaining_chunk_count_ > 0);
+    remaining_chunk_count_--;
+    lock.unlock();
+    cvr_.notify_all();
+  }
+
+  /// Push preloaded chunks to batch queue. Called from the ChunkDataSet worker
+  /// threads.
+  void add_chunk_data(size_t index, UnwrappedBatchType data) {
```

Comment:
`index` is completely unused. Was this supposed to do some ordering, which is not implemented?